File: reference_example/android/bridge/src/androidTest/java/ipfs/gomobile/android/requestIPFSTests.java
package ipfs.gomobile.android;

import android.content.Context;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.Timeout;
import org.junit.runner.RunWith;

import static org.junit.Assert.*;

/**
* Instrumented test, which will execute on an Android device.
*
* @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
*/
@RunWith(AndroidJUnit4.class)
public class requestIPFSTests {
    private IPFS ipfs;

    @Rule
    public Timeout globalTimeout = Timeout.seconds(600);

    @Before
    public void setup() throws Exception {
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        ipfs = new IPFS(appContext);
        ipfs.start();
    }

    @Test
    public void testDNSRequest() throws Exception {
        String domain = "website.ipfs.io";

        JSONObject resolveResp = ipfs.newRequest("resolve")
                .withArgument("/ipns/" + domain)
                .sendToJSONList()
                .get(0);
        JSONObject dnsResp = ipfs.newRequest("dns")
                .withArgument(domain)
                .sendToJSONList()
                .get(0);

        String resolvePath = resolveResp.getString("Path");
        String dnsPath = dnsResp.getString("Path");

        assertEquals(
            "resolve and dns request should return the same result",
            resolvePath,
            dnsPath
        );

        assertEquals(
            "response should start with \"/ipfs/\"",
            dnsPath.substring(0, 6),
            "/ipfs/"
        );
    }

    @Test
    public void testCatFile() throws Exception {
        byte[] latestRaw = ipfs.newRequest("cat")
                .withArgument("/ipns/xkcd.hacdias.com/latest/info.json")
                .send();

        try {
            new JSONObject(new String(latestRaw));
        } catch (JSONException e) {
            fail("error while parsing fetched JSON: " + new String(latestRaw));
        }
    }
}

File: ios/Bridge/GomobileIPFSTests/RequestIPFSTests.swift
//
//  RequestIPFSTests.swift
//  GomobileIPFSTests
//
//  Created by Antoine Eddi on 4/15/20.
//  Copyright © 2020 Antoine Eddi. All rights reserved.
//

import XCTest
@testable import GomobileIPFS

class RequestIPFSTests: XCTestCase {

    private var ipfs: IPFS!

    override func setUp() {
        do {
            ipfs = try IPFS()
            try ipfs!.start()
        } catch _ {
            XCTFail("IPFS initialization failed")
        }
    }

    func testDNSRequest() throws {
        let domain = "website.ipfs.io"

        let resolveResp = try ipfs.newRequest("resolve")
            .with(argument: "/ipns/\(domain)")
            .sendToDict()
        let dnsResp = try ipfs.newRequest("dns")
            .with(argument: domain)
            .sendToDict()

        guard let resolvePath = resolveResp["Path"] as? String else {
            XCTFail("error while casting value associated to \"Path\" key")
            return
        }
        guard let dnsPath = dnsResp["Path"] as? String else {
            XCTFail("error while casting value associated to \"Path\" key")
            return
        }
        let index = dnsPath.index(dnsPath.startIndex, offsetBy: 6)

        XCTAssertEqual(
            resolvePath,
            dnsPath,
            "resolve and dns request should return the same result"
        )

        XCTAssertEqual(
            dnsPath[..<index],
            "/ipfs/",
            "response should start with \"/ipfs/\""
        )
    }

    func testCatFile() throws {
        let latestRaw = try ipfs.newRequest("cat")
            .with(argument: "/ipns/xkcd.hacdias.com/latest/info.json")
            .send()

        do {
            try JSONSerialization.jsonObject(with: latestRaw, options: [])
        } catch _ {
            XCTFail("error while parsing fetched JSON:  \(String(decoding: latestRaw, as: UTF8.self))")
        }
    }
}
